Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,447 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+import Darwin
+import CNIOLinux
+import CoreFoundation
+
+
+/// `Error` that may be thrown if we fail to create a `IPAddress`
+public enum IPAddressError: Error {
+    /// Given string input is not supported IP Address
+    case failedToParseIPString(String)
+    /// Given string input is not supported IP Address
+    case bytesArrayHasWrongLength(Int)
+}
+
+public typealias IPv4BytesTuple = (UInt8, UInt8, UInt8, UInt8)
+public typealias IPv6BytesTuple = (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
+
+/// Represent the bytes for an `IPv4Address`.
+public struct IPv4Bytes {
+    private let ipv4BytesTuple: IPv4BytesTuple
+    
+    public var bytes: IPv4BytesTuple {
+        return self.ipv4BytesTuple
+    }
+    
+    init(_ bytes: IPv4BytesTuple) {
+        self.ipv4BytesTuple = bytes
+    }
+}
+
+/// Represent the bytes for an `IPv6Address.
+public struct IPv6Bytes {
+    private let ipv6BytesTuple: IPv6BytesTuple
+    
+    public var bytes: IPv6BytesTuple {
+        return self.ipv6BytesTuple
+    }
+    
+    init(_ bytes: IPv6BytesTuple) {
+        self.ipv6BytesTuple = bytes
+    }
+}
+
+/// Represent a single `IPAddress`.
+public enum IPAddress {
+    /// A single IPv4 address for `IPAddress`.
+    public struct IPv4Address {
+        /// The bytes storing the address of the IPv4 address.
+        let address: IPv4Bytes
+        
+        /// Get the `IPv4Address` as a string.
+        var ipAddress: String {
```

Comment:
This should probably be moved into a `var description` on a `CustomStringConvertible` conformance.